janusgraph-core/src/main/java/org/janusgraph/diskstorage/util/MetricInstrumentedIndexProvider.java

```diff
@@ -0,0 +1,174 @@
+        return runWithMetrics((BaseTransactionConfigurable) tx, M_RAW_QUERY, () -> indexProvider.query(query, information, tx));
+    }
+
+    @Override
+    public Long totals(final RawQuery query, final KeyInformation.IndexRetriever information, final BaseTransaction tx) throws BackendException {
+        return runWithMetrics((BaseTransactionConfigurable) tx, M_TOTALS, () -> indexProvider.totals(query, information, tx));
+    }
+
+    @Override
+    public BaseTransactionConfigurable beginTransaction(final BaseTransactionConfig config) throws BackendException {
+        return indexProvider.beginTransaction(config);
+    }
+
+    @Override
+    public void close() throws BackendException {
+        indexProvider.close();
+    }
+
+    @Override
+    public void clearStorage() throws BackendException {
+        indexProvider.clearStorage();
+    }
+
+    @Override
+    public boolean exists() throws BackendException {
+        return indexProvider.exists();
+    }
+
+    @Override
+    public boolean supports(final KeyInformation information, final JanusGraphPredicate janusgraphPredicate) {
+        return indexProvider.supports(information, janusgraphPredicate);
+    }
+
+    @Override
+    public boolean supports(final KeyInformation information) {
+        return indexProvider.supports(information);
+    }
+
+    @Override
+    public String mapKey2Field(final String key, final KeyInformation information) {
+        return indexProvider.mapKey2Field(key, information);
+    }
+
+    @Override
+    public IndexFeatures getFeatures() {
+        return indexProvider.getFeatures();
+    }
+
+    void runWithMetrics(BaseTransactionConfigurable tx, String name, StorageRunnable impl) throws BackendException {
+        if (null == prefix) {
+            impl.run();
+        }
+
+        Preconditions.checkNotNull(name);
+        Preconditions.checkNotNull(impl);
+
+        String groupName = tx.getConfiguration().getGroupName();
+
+        metricManager.getCounter(groupName, prefix, name, M_CALLS).inc();
+        final Timer.Context tc = metricManager.getTimer(groupName, prefix, name, M_TIME).time();
+        try {
+            impl.run();
+        } catch (RuntimeException e) {
+            metricManager.getCounter(groupName, prefix, name, M_EXCEPTIONS).inc();
+            throw e;
+        } finally {
+            tc.stop();
+        }
+    }
+
+    <T> T runWithMetrics(BaseTransactionConfigurable tx, String name, StorageCallable<T> impl) throws BackendException {
+        if (null == prefix) {
+            return impl.call();
+        }
+
+        Preconditions.checkNotNull(name);
+        Preconditions.checkNotNull(impl);
+
+        String groupName = tx.getConfiguration().getGroupName();
+
+        metricManager.getCounter(groupName,prefix, name, M_CALLS).inc();
+        final Timer.Context tc = metricManager.getTimer(groupName, prefix, name, M_TIME).time();
```

Comment:
I think it would be good to move this call into a separate small methods to reduce duplicate code.
Something like:
```
private Timer.Context incrementCallsAndReturnTimerContext(String groupName, String name){
    metricManager.getCounter(groupName, prefix, name, M_CALLS).inc();
    return metricManager.getTimer(groupName, prefix, name, M_TIME).time();
}

private void incrementExceptions(String groupName, String name){
    metricManager.getCounter(groupName, prefix, name, M_EXCEPTIONS).inc();
}
```

Reply:
ok